invisible_cities/io/mcinfo_io.py

```diff
@@ -168,6 +172,37 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
         self.generator_table.flush()
 
 
+def copy_mc_info(files_in, h5out, event_list):
+    """
+    Copies MC info for selected events to an output file.
+    files_in :   List[str]
+                 List with the name of the input files.
+    file_out   : str
+                 The name of the output file.
+    event_list : List[int]
+                 List of events (event_id) to be copied.
+    """
+
+    writer = mc_info_writer(h5out)
+
+    for f in files_in:
+        try:
+            with tb.open_file(f, "r") as h5in:
+                run_number = get_run_number(h5in)
+                mc_info    = get_mc_info_safe(h5in, run_number)
+                for n in event_list[:]:
+                    try:
+                        writer(mctables=mc_info, evt_number=n)
+                        # We succeeded copying the event. Remove its id
+                        # from the event list
+                        event_list.remove(n)
+                    except IndexError:
+                        # The event id was not found in the current file
+                        continue
```

Comment:
A few things:
- There is nothing to ensure that `event_list` is empty at the end of all loops... I wonder if we can make sure of that.
- Maybe we can get rid of this `try/except` clause by using `set`s. If you take the event numbers of the input file and make the intersection with `event_list` you would only run over the event numbers in that file without having to check them all.
- I think we don't have any protection against repeated event numbers in the writer, but it's something to think about (not for this PR, though)

Reply:
> Maybe we can get rid of this try/except clause by using sets. If you take the event numbers of the input file and make the intersection with event_list you would only run over the event numbers in that file without having to check them all.

This function will be modified in #693, so I dont think we need to fine tune it too much. However this implementation has 2 issues:

1. it sneakily modifies event_list
2. in case event numbers are the same across different input files just the first occurrence (file, event number) will be written.  I am not sure how big of an issue it is cause we do not have any protection anywhere against repeated event numbers in our readers and writers ...

Reply:
I can fix somewhat the _sneaky_ behavior of the function. However, there were several things about event_list I wasn't sure I could assume; namely:
- Do event ids come in order?
- Is that order correlated with the files_in order?
- Can event ids repeat?
I understand that the answer to the first two is 'no', and that we really don't know the answer of the third one, correct?

Reply:
> Do event ids come in order?

Not necessarily

> Is that order correlated with the `files_in` order?

They will keep the order in which they are written within the same file, and the order of the input files is kept. That is if you have `file_1` with events `[2, 6, 4]` and `file_2` with events `[1, 3, 5]`
you will receive `[2, 6, 4, 1, 3, 5]`

> Can event ids repeat?

They can, in theory. It shouldn't happen, but there is nothing preventing it. For instance if you run over two different MC productions at the same time, the event numbers will repeat unless you take care of it explicitly. It can also happen if someone doesn't take care of organizing the event number in a `nexus` run that is split into many jobs. So far we haven't had issues, but I think it can happen in #641.

Reply:
So far, we have assumed that event IDs don't repeat, because we were using dictionaries in the event data model, correct me if I'm wrong. I'm not sure this is the case anymore, so we may want to discuss this point. Not to repeat event IDs is particularly cumbersome when we do event mixing in the background model simulations, so, if we can drop this requirement, it may be a simplification.

Reply:
Uhh, this is getting complicated... There are couple of options:

- we assume events ids and files_in order is correlated (in that case removing the event_id from the list should work well even in the case of duplicates)
- we write all combinations of files_in, events_ids that are possible (probably not very useful)
- we move this function to components and not worry about general usage 

What is your opinion @gonzaponte ?

Reply:
Now that i think of it the safest would be to have a dictionary `{file_in : event_list} `, that way there can not be confusion. However we will not be able to read this file since as @paolafer  said we do expect to have unique IDs at the reader level...

Reply:
How about making this function for a single file and expand it for many in components? This would make the utility available in general (which is probably useful) while keeping these intricacies away from the base functions. I don't think we need to deal with repeated events or the ordering here. My feeling is that we should assume that order doesn't matter and that there shouldn't be repeated events for practical purposes. In case, we can add a tool that ensures that somewhere.

Reply:
> How about making this function for a single file and expand it for many in components? This would make the utility available in general (which is probably useful) while keeping these intricacies away from the base functions. I don't think we need to deal with repeated events or the ordering here. My feeling is that we should assume that order doesn't matter and that there shouldn't be repeated events for practical purposes. In case, we can add a tool that ensures that somewhere.

But how do I correlate files with selected events then? The main loop in push is reading all input files together.

Reply:
I propose to do the following:

In `mcinfo_io.copy_mc_info`
- Make this function take `h5in`,`h5out` and `which_events` (unofficial names)
- If `which_events` is `None` (probably default), read the event numbers from the input file and assign it to `which_events`. This is a detail that doesn't affect the current usage, but that it would be good to have.
- Assume that each event in `which_events` is stored in the file.
- For each event in `which_events`, copy it.

In `components.copy_mc_info` (unofficial name):
- take `filenames_in`, `h5out` and `all_filtered_events`
- for each filename
  - open the file and call it `h5in` (unofficial name)
  - read the event numbers in that file and make the intersection with `all_filtered_events`. Let's call that `filtered_events_in_file`
  - call `mc_info_io.copy_mc_info(h5in, h5out, filtered_events_in_file)`

With this:
- Duplicated event numbers in different files are written
- The order of the files is irrelevant
- The order of the events within a file matters (as it has mattered so far), but hopefully this will be addressed in the future

Reply:
Sounds good, @gonzaponte.

Reply:
Also dont forget to change get_mc_info_safe to get_mc_info directly!